packages/multichain-account-service/src/MultichainAccountWallet.ts

```diff
@@ -326,4 +326,101 @@ export class MultichainAccountWallet<
       await group.align();
     }
   }
+
+  /**
+   * Discover and create accounts for all providers.
+   *
+   * NOTE: This method should only be called on a newly created wallet.
+   *
+   * @param opts - The options for the discovery and creation of accounts.
+   * @param opts.skippedProviders - The providers to skip.
+   * @returns The accounts for each provider.
+   */
+  async discoverAndCreateAccounts({
+    skippedProviders = [],
+  }: {
+    skippedProviders: AccountProviderType[];
+  }): Promise<{
+    [AccountProviderType]: Bip44Account<KeyringAccount>[];
+  }> {
+    const providers = this.#providers.filter(
+      (provider) => !skippedProviders.includes(provider.providerType),
+    );
+
+    const results: Record<AccountProviderType, Bip44Account<KeyringAccount>[]> =
+      {};
+    const nextIndex = new Map<Bip44Provider, number>(
+      providers.map((p) => [p, 0]),
+    );
+    const stopped = new Set<Bip44Provider>();
+    const inFlight = new Map<Bip44Provider, Promise<void>>();
+
+    let high = 0;
+
+    const schedule = (p: Bip44Provider, index: number) => {
+      const run = (async () => {
+        try {
+          const accounts = await p.discoverAndCreateAccounts({
+            entropySource: this.#entropySource,
+            groupIndex: index,
+          });
+
+          inFlight.delete(p);
+
+          if (accounts.length > 0) {
+            stopped.add(p);
+          } else {
+            results[p.providerType] = [
+              ...(results[p.providerType] ?? []),
+              ...accounts,
+            ];
+
+            const next = index + 1;
+            nextIndex.set(p, next);
+            if (next > high) {
+              high = next;
+
+              for (const q of providers) {
+                if (
+                  !stopped.has(q) &&
+                  !inFlight.has(q) &&
+                  (nextIndex.get(q) ?? 0) < high
+                ) {
+                  schedule(q, high);
+                }
+              }
+            }
+
+            if (!stopped.has(p)) {
+              const target = Math.max(high, nextIndex.get(p) ?? 0);
+              schedule(p, target);
+            }
+          }
+        } catch {
+          inFlight.delete(p);
+          stopped.add(p);
+        }
+      })();
+
+      inFlight.set(p, run);
+    };
+
+    for (const p of providers) {
+      schedule(p, 0);
+    }
+
+    while (inFlight.size > 0) {
+      await Promise.race(inFlight.values());
+    }
```

Comment:
I wonder if this is truly concurrent-safe here 🤔 Since we're using `.values()` we will be using iterators, and the first "scheduled job" might alter the `inFlight` map. So maybe we should use `Array.from(inFlight.values())` just in case?